File: bishopTests/src/bishopTests/BitBoardCombinatorTest.java
package bishopTests;

import bishop.base.BitBoard;
import bishop.base.BitBoardCombinator;
import bishop.base.Square;
import org.junit.Assert;
import org.junit.Test;

public class BitBoardCombinatorTest {
	private int[][] TESTED_MASKS = {
			{Square.A1, Square.B1, Square.H8},
			{Square.B5, Square.C5, Square.D6, Square.E6}
	};

	@Test
	public void test() {
		for (int[] squares: TESTED_MASKS) {
			final BitBoardCombinator combinator = new BitBoardCombinator(BitBoard.of(squares));

			for (int i = 0; i < (1 << squares.length); i++) {
				long expectedCombination = 0;

				for (int j = 0; j < squares.length; j++) {
					final int square = squares[j];

					expectedCombination |= ((i >>> j) & 0x01L) << square;
				}

				Assert.assertTrue(combinator.hasNextCombination());
				Assert.assertEquals(expectedCombination, combinator.getNextCombination());
			}

			Assert.assertFalse(combinator.hasNextCombination());
		}
	}
}

File: native/tests/bishop/bit_board_combinator_test.cpp


#include "../../bishop/base/bit_board_combinator.h"
#include "../test_base/test_base.h"

#include <vector>

using namespace bishop::base;


std::vector<std::vector<Square::Type>> TESTED_MASKS = {
	{Square::A1, Square::B1, Square::H8},
	{Square::B5, Square::C5, Square::D6, Square::E6}
};

TEST (test) {
	for (const std::vector<Square::Type> &squares: TESTED_MASKS) {
		BitBoardCombinator combinator(BitBoard::fromRange(squares.begin(), squares.end()));

		for (unsigned int i = 0; i < (1ULL << squares.size()); i++) {
			BitBoard::Type expectedCombination = BitBoard::EMPTY;

			for (size_t j = 0; j < squares.size(); j++) {
				const Square::Type square = squares[j];

				expectedCombination |= (BitBoard::Type) ((i >> j) & 0x01L) << square;
			}

			assertTrue(combinator.hasNextCombination());
			assertEquals(expectedCombination, combinator.getNextCombination());
		}

		assertFalse(combinator.hasNextCombination());
	}
}

TEST_RUNNER_BEGIN
	RUN_TEST(test);
TEST_RUNNER_END
